transport-native-kqueue/src/main/c/netty_kqueue_bsdsocket.c

```diff
@@ -83,6 +83,57 @@ static jlong netty_kqueue_bsdsocket_sendFile(JNIEnv* env, jclass clazz, jint soc
     return res < 0 ? -err : 0;
 }
 
+static jint netty_kqueue_bsdsocket_connectx(JNIEnv* env,
+        jint socketFd,
+        jint socketInterface,
+        jboolean sourceIPv6, jbyteArray sourceAddress, jint sourceScopeId, jint sourcePort,
+        jboolean destinationIPv6, jbyteArray destinationAddress, jint destinationScopeId, jint destinationPort,
+        jint flags,
+        jlong iovAddress, jint iovCount, jint iovDataLength) {
+#ifdef __APPLE__ // connectx(2) is only defined on Darwin.
+    sa_endpoints_t endpoints;
+    endpoints.sae_srcif = (unsigned int) socketInterface;
+    endpoints.sae_srcaddr = NULL;
+    endpoints.sae_srcaddrlen = 0;
+    endpoints.sae_dstaddr = NULL;
+    endpoints.sae_dstaddrlen = 0;
+
+    struct sockaddr_storage srcaddr;
+    socklen_t srcaddrlen;
+    struct sockaddr_storage dstaddr;
+    socklen_t dstaddrlen;
+
+    if (NULL != sourceAddress) {
+        if (-1 == netty_unix_socket_initSockaddr(env,
+                sourceIPv6, sourceAddress, sourceScopeId, sourcePort, &srcaddr, &srcaddrlen)) {
+            return -1;
+        }
+        endpoints.sae_srcaddr = (const struct sockaddr*) &srcaddr;
```

Comment:
instead of declaring independent memory on the stack, can you just use `endpoints.sae_srcaddr` directly? this method is self contained but we would run into hard to track memory access issues if the lifetime of `endpoints` escaped this method.

(same comment below for dst address)

Reply:
If they are not on the stack, then they are on the heap. The `sae_srcaddr` is a pointer field. The `connectx` call is blocking, so I can't imagine that the `endpoints` would outlive it. If it did, we would have no idea about when it would be safe to free the memory.

Reply:
But at that point, `endpoints.sae_srcaddr` is a pointer to `NULL`. The `sa_endpoints_t` does not have room for the addresses themselves, it's `sae_srcaddr` and `sae_dstaddr` fields are pointers.

Reply:
For reference, `sa_endpoints_t` is defined like this:

```c
     typedef struct sa_endpoints {
             unsigned int     sae_srcif;      /* optional source interface   */
             struct sockaddr *sae_srcaddr;    /* optional source address     */
             socklen_t        sae_srcaddrlen; /* size of source address      */
             struct sockaddr *sae_dstaddr;    /* destination address         */
             socklen_t        sae_dstaddrlen; /* size of destination address */
     }sa_endpoints_t;
```

Reply:
👍 thx for clarifying! feel free to disregard.